from abc import ABC
from datetime import datetime
from janis_core import (
    CommandTool,
    ToolInput,
    ToolOutput,
    File,
    Boolean,
    String,
    Int,
    InputSelector,
    Filename,
    ToolMetadata,
)

from ..gatk4toolbase import Gatk4ToolBase


class Gatk4SelectVariantsBase(Gatk4ToolBase, ABC):
    def friendly_name(self) -> str:
        return "GATK4: SelectVariants"

    @staticmethod
    def tool() -> str:
        return "Gatk4SelectVariants"

    @classmethod
    def gatk_command(cls):
        return "SelectVariants"

    def inputs(self):
        return [
            ToolInput(
                tag="outputFilename",
                input_type=Filename(optional=True),
                prefix="-O",
                separate_value_from_prefix=True,
                doc="(--output) Required.",
            ),
            ToolInput(
                tag="variants",
                input_type=VcfTabix(optional=True),
                prefix="-V",
                separate_value_from_prefix=True,
                doc="(--variant) A VCF file containing variants Required.",
            ),
            ToolInput(
                tag="addOutputSamProgramRecord",
                input_type=Boolean(optional=True),
                prefix="-add-output-sam-program-record:Boolean",
                separate_value_from_prefix=True,
                doc="(--add-output-sam-program-record)  If true, adds a PG tag to created SAM/BAM/CRAM files.  Default value: true. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="addOutputVcfCommandLine",
                input_type=Boolean(optional=True),
                prefix="-add-output-vcf-command-line",
                separate_value_from_prefix=True,
                doc="(--add-output-vcf-command-line)  If true, adds a command line header line to created VCF files.  Default value: true. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="arguments_file",
                input_type=File(optional=True),
                prefix="--arguments_file",
                separate_value_from_prefix=True,
                doc="read one or more arguments files and add them to the command line This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="cloudIndexPrefetchBuffer",
                input_type=Int(optional=True),
                prefix="--cloud-index-prefetch-buffer",
                separate_value_from_prefix=True,
                doc="(-CIPB:Integer)  Size of the cloud-only prefetch buffer (in MB; 0 to disable). Defaults to cloudPrefetchBuffer if unset.  Default value: -1. ",
            ),
            ToolInput(
                tag="cloudPrefetchBuffer",
                input_type=Int(optional=True),
                prefix="--cloud-prefetch-buffer",
                separate_value_from_prefix=True,
                doc="(-CPB:Integer)  Size of the cloud-only prefetch buffer (in MB; 0 to disable).  Default value: 40. ",
            ),
            ToolInput(
                tag="conc",
                input_type=String(optional=True),
                prefix="-conc",
                separate_value_from_prefix=True,
                doc="(--concordance)  Output variants also called in this comparison track  Default value: null. ",
            ),
            ToolInput(
                tag="createOutputBamIndex",
                input_type=Boolean(optional=True),
                prefix="--create-output-bam-index",
                doc="(-OBI)  If true, create a BAM/CRAM index when writing a coordinate-sorted BAM/CRAM file.  Default value: true. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="createOutputBamMd5",
                input_type=Boolean(optional=True),
                prefix="--create-output-bam-md5",
                doc="(-OBM)  If true, create a MD5 digest for any BAM/SAM/CRAM file created  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="createOutputvariantIndex",
                input_type=Boolean(optional=True),
                prefix="--create-output-variant-index",
                default=True,
                doc="(-OVI)  If true, create a VCF index when writing a coordinate-sorted VCF file.  Default value: true. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="createOutputvariantMd5",
                input_type=Boolean(optional=True),
                prefix="--create-output-variant-md5",
                doc="(-OVM)  If true, create a a MD5 digest any VCF file created.  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="disableBamIndexCaching",
                input_type=Boolean(optional=True),
                prefix="--disable-bam-index-caching",
                doc="(-DBIC:Boolean)  If true, don't cache bam indexes, this will reduce memory requirements but may harm performance if many intervals are specified.  Caching is automatically disabled if there are no intervals specified.  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="disableReadFilter",
                input_type=String(optional=True),
                prefix="--disable-read-filter",
                separate_value_from_prefix=True,
                doc="(-DF)  Read filters to be disabled before analysis  This argument may be specified 0 or more times. Default value: null. Possible Values: {WellformedReadFilter}",
            ),
            ToolInput(
                tag="disableSequenceDictionaryValidation",
                input_type=Boolean(optional=True),
                prefix="-disable-sequence-dictionary-validation",
                doc="(--disable-sequence-dictionary-validation)  If specified, do not check the sequence dictionaries from our inputs for compatibility. Use at your own risk!  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="disc",
                input_type=String(optional=True),
                prefix="-disc",
                separate_value_from_prefix=True,
                doc="(--discordance)  Output variants not called in this comparison track  Default value: null. ",
            ),
            ToolInput(
                tag="dropGenotypeAnnotation",
                input_type=String(optional=True),
                prefix="--drop-genotype-annotation",
                separate_value_from_prefix=True,
                doc="(-DGA:String)  Genotype annotations to drop from output vcf.  Annotations to be dropped are specified by their key.  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="dropInfoAnnotation",
                input_type=String(optional=True),
                prefix="--drop-info-annotation",
                separate_value_from_prefix=True,
                doc="(-DA:String)  Info annotations to drop from output vcf.  Annotations to be dropped are specified by their key.  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="excludeFiltered",
                input_type=Boolean(optional=True),
                prefix="--exclude-filtered",
                doc="Don't include filtered sites Default value: false. Possible values: {true, false}",
            ),
            ToolInput(
                tag="xlIds",
                input_type=String(optional=True),
                prefix="-xl-ids",
                separate_value_from_prefix=True,
                doc="(--exclude-ids) List of variant rsIDs to exclude This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="excludeIntervals",
                input_type=String(optional=True),
                prefix="--exclude-intervals",
                separate_value_from_prefix=True,
                doc="(-XL) This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="excludeNonvariants",
                input_type=String(optional=True),
                prefix="--exclude-non-variants",
                doc="Default value: false. Possible values: {true, false}",
            ),
            ToolInput(
                tag="excludeSampleExpressions",
                input_type=String(optional=True),
                prefix="--exclude-sample-expressions",
                separate_value_from_prefix=True,
                doc="(-xl-se:String)  List of sample expressions to exclude  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="excludeSampleName",
                input_type=String(optional=True),
                prefix="--exclude-sample-name",
                separate_value_from_prefix=True,
                doc="(-xl-sn:String)  Exclude genotypes from this sample  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="gatkConfigFile",
                input_type=File(optional=True),
                prefix="--gatk-config-file",
                separate_value_from_prefix=True,
                doc="A configuration file to use with the GATK. Default value: null.",
            ),
            ToolInput(
                tag="gcsRetries",
                input_type=Int(optional=True),
                prefix="-gcs-retries",
                separate_value_from_prefix=True,
                doc="(--gcs-max-retries)  If the GCS bucket channel errors out, how many times it will attempt to re-initiate the connection  Default value: 20. ",
            ),
            ToolInput(
                tag="gcsProjectForRequesterPays",
                input_type=String(optional=True),
                prefix="--gcs-project-for-requester-pays",
                separate_value_from_prefix=True,
                doc=" Project to bill when accessing requester pays buckets. If unset, these buckets cannot be accessed.  Default value: . ",
            ),
            ToolInput(
                tag="help",
                input_type=Boolean(optional=True),
                prefix="-h",
                doc="(--help) display the help message Default value: false. Possible values: {true, false}",
            ),
            ToolInput(
                tag="bam",
                input_type=BamBai(optional=True),
                prefix="-I",
                separate_value_from_prefix=True,
                doc="(--input) BAM/SAM/CRAM file containing reads This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="intervalExclusionPadding",
                input_type=Int(optional=True),
                prefix="--interval-exclusion-padding",
                separate_value_from_prefix=True,
                doc="(-ixp:Integer)  Amount of padding (in bp) to add to each interval you are excluding.  Default value: 0. ",
            ),
            ToolInput(
                tag="imr",
                input_type=String(optional=True),
                prefix="-imr",
                separate_value_from_prefix=True,
                doc="(--interval-merging-rule)  Interval merging rule for abutting intervals  Default value: ALL. Possible values: {ALL, OVERLAPPING_ONLY} ",
            ),
            ToolInput(
                tag="ip",
                input_type=Int(optional=True),
                prefix="-ip",
                separate_value_from_prefix=True,
                doc="(--interval-padding) Default value: 0.",
            ),
            ToolInput(
                tag="isr",
                input_type=String(optional=True),
                prefix="-isr",
                separate_value_from_prefix=True,
                doc="(--interval-set-rule)  Set merging approach to use for combining interval inputs  Default value: UNION. Possible values: {UNION, INTERSECTION} ",
            ),
            ToolInput(
                tag="intervals",
                input_type=String(optional=True),
                prefix="--intervals",
                separate_value_from_prefix=True,
                doc="(-L:String) One or more genomic intervals over which to operate This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="invertMendelianViolation",
                input_type=Boolean(optional=True),
                prefix="--invert-mendelian-violation",
                doc=" Output non-mendelian violation sites only  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="invertSelect",
                input_type=Boolean(optional=True),
                prefix="-invert-select",
                doc="(--invertSelect)  Invert the selection criteria for -select  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="ids",
                input_type=String(optional=True),
                prefix="-ids",
                separate_value_from_prefix=True,
                doc="(--keep-ids) List of variant rsIDs to select This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="keepOriginalAc",
                input_type=Boolean(optional=True),
                prefix="--keep-original-ac",
                doc="Store the original AC, AF, and AN values after subsetting Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="keepOriginalDp",
                input_type=Boolean(optional=True),
                prefix="--keep-original-dp",
                separate_value_from_prefix=True,
                doc="Store the original DP value after subsetting Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="le",
                input_type=Boolean(optional=True),
                prefix="-LE",
                separate_value_from_prefix=True,
                doc="(--lenient) Lenient processing of VCF files Default value: false. Possible values: {true, false}",
            ),
            ToolInput(
                tag="maxFilteredGenotypes",
                input_type=Int(optional=True),
                prefix="--max-filtered-genotypes",
                separate_value_from_prefix=True,
                doc="Maximum number of samples filtered at the genotype level  Default value: 2147483647. ",
            ),
            ToolInput(
                tag="maxFractionFilteredGenotypes",
                input_type=Double(optional=True),
                prefix="--max-fraction-filtered-genotypes",
                separate_value_from_prefix=True,
                doc=" Maximum fraction of samples filtered at the genotype level  Default value: 1.0. ",
            ),
            ToolInput(
                tag="maxIndelSize",
                input_type=Int(optional=True),
                prefix="--max-indel-size",
                separate_value_from_prefix=True,
                doc="Maximum size of indels to include Default value: 2147483647.",
            ),
            ToolInput(
                tag="maxNocallFraction",
                input_type=Double(optional=True),
                prefix="--max-nocall-fraction",
                separate_value_from_prefix=True,
                doc="Maximum fraction of samples with no-call genotypes Default value: 1.0.",
            ),
            ToolInput(
                tag="maxNocallNumber",
                input_type=Int(optional=True),
                prefix="--max-nocall-number",
                separate_value_from_prefix=True,
                doc="Maximum number of samples with no-call genotypes Default value: 2147483647.",
            ),
            ToolInput(
                tag="mendelianViolation",
                input_type=Boolean(optional=True),
                prefix="--mendelian-violation",
                separate_value_from_prefix=True,
                doc="Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="mendelianViolationQualThreshold",
                input_type=Double(optional=True),
                prefix="--mendelian-violation-qual-threshold",
                separate_value_from_prefix=True,
                doc=" Minimum GQ score for each trio member to accept a site as a violation  Default value: 0.0.",
            ),
            ToolInput(
                tag="minFilteredGenotypes",
                input_type=Int(optional=True),
                prefix="--min-filtered-genotypes",
                separate_value_from_prefix=True,
                doc=" Minimum number of samples filtered at the genotype level  Default value: 0. ",
            ),
            ToolInput(
                tag="minFractionFilteredGenotypes",
                input_type=Double(optional=True),
                prefix="--min-fraction-filtered-genotypes",
                separate_value_from_prefix=True,
                doc=" Maximum fraction of samples filtered at the genotype level  Default value: 0.0. ",
            ),
            ToolInput(
                tag="minIndelSize",
                input_type=Int(optional=True),
                prefix="--min-indel-size",
                separate_value_from_prefix=True,
                doc="Minimum size of indels to include Default value: 0.",
            ),
            ToolInput(
                tag="pedigree",
                input_type=File(optional=True),
                prefix="--pedigree",
                separate_value_from_prefix=True,
                doc="(-ped:File) Pedigree file Default value: null.",
            ),
            ToolInput(
                tag="preserveAlleles",
                input_type=Boolean(optional=True),
                prefix="--preserve-alleles",
                separate_value_from_prefix=True,
                doc="Preserve original alleles, do not trim Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="quiet",
                input_type=Boolean(optional=True),
                prefix="--QUIET",
                separate_value_from_prefix=True,
                doc="Whether to suppress job-summary info on System.err. Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="readFilter",
                input_type=String(optional=True),
                prefix="--read-filter",
                separate_value_from_prefix=True,
                doc="(-RF:String) Read filters to be applied before analysis This argument may be specified 0 or more times. Default value: null. Possible Values: {AlignmentAgreesWithHeaderReadFilter, AllowAllReadsReadFilter, AmbiguousBaseReadFilter, CigarContainsNoNOperator, FirstOfPairReadFilter, FragmentLengthReadFilter, GoodCigarReadFilter, HasReadGroupReadFilter, IntervalOverlapReadFilter, LibraryReadFilter, MappedReadFilter, MappingQualityAvailableReadFilter, MappingQualityNotZeroReadFilter, MappingQualityReadFilter, MatchingBasesAndQualsReadFilter, MateDifferentStrandReadFilter, MateOnSameContigOrNoMappedMateReadFilter, MateUnmappedAndUnmappedReadFilter, MetricsReadFilter, NonChimericOriginalAlignmentReadFilter, NonZeroFragmentLengthReadFilter, NonZeroReferenceLengthAlignmentReadFilter, NotDuplicateReadFilter, NotOpticalDuplicateReadFilter, NotSecondaryAlignmentReadFilter, NotSupplementaryAlignmentReadFilter, OverclippedReadFilter, PairedReadFilter, PassesVendorQualityCheckReadFilter, PlatformReadFilter, PlatformUnitReadFilter, PrimaryLineReadFilter, ProperlyPairedReadFilter, ReadGroupBlackListReadFilter, ReadGroupReadFilter, ReadLengthEqualsCigarLengthReadFilter, ReadLengthReadFilter, ReadNameReadFilter, ReadStrandFilter, SampleReadFilter, SecondOfPairReadFilter, SeqIsStoredReadFilter, SoftClippedReadFilter, ValidAlignmentEndReadFilter, ValidAlignmentStartReadFilter, WellformedReadFilter}",
            ),
            ToolInput(
                tag="readIndex",
                input_type=File(optional=True),
                prefix="-read-index",
                separate_value_from_prefix=True,
                doc="(--read-index)  Indices to use for the read inputs. If specified, an index must be provided for every read input and in the same order as the read inputs. If this argument is not specified, the path to the index for each input will be inferred automatically.  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="readValidationStringency",
                input_type=String(optional=True),
                prefix="--read-validation-stringency",
                separate_value_from_prefix=True,
                doc="(-VS:ValidationStringency)  Validation stringency for all SAM/BAM/CRAM/SRA files read by this program.  The default stringency value SILENT can improve performance when processing a BAM file in which variable-length data (read, qualities, tags) do not otherwise need to be decoded.  Default value: SILENT. Possible values: {STRICT, LENIENT, SILENT} ",
            ),
            ToolInput(
                tag="reference",
                input_type=FastaWithDict(optional=True),
                prefix="--reference",
                separate_value_from_prefix=True,
                doc="(-R:String) Reference sequence Default value: null.",
            ),
            ToolInput(
                tag="removeFractionGenotypes",
                input_type=Double(optional=True),
                prefix="--remove-fraction-genotypes",
                separate_value_from_prefix=True,
                doc=" Select a fraction of genotypes at random from the input and sets them to no-call  Default value: 0.0. ",
            ),
            ToolInput(
                tag="removeUnusedAlternates",
                input_type=Boolean(optional=True),
                prefix="--remove-unused-alternates",
                separate_value_from_prefix=True,
                doc=" Remove alternate alleles not present in any genotypes  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="restrictAllelesTo",
                input_type=String(optional=True),
                prefix="--restrict-alleles-to",
                separate_value_from_prefix=True,
                doc=" Select only variants of a particular allelicity  Default value: ALL. Possible values: {ALL, BIALLELIC, MULTIALLELIC} ",
            ),
            ToolInput(
                tag="sampleExpressions",
                input_type=String(optional=True),
                prefix="--sample-expressions",
                separate_value_from_prefix=True,
                doc="(-se:String)  Regular expression to select multiple samples  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="sampleName",
                input_type=String(optional=True),
                prefix="--sample-name",
                separate_value_from_prefix=True,
                doc="(-sn:String) Include genotypes from this sample This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="secondsBetweenProgressUpdates",
                input_type=Double(optional=True),
                prefix="-seconds-between-progress-updates",
                separate_value_from_prefix=True,
                doc="(--seconds-between-progress-updates)  Output traversal statistics every time this many seconds elapse  Default value: 10.0. ",
            ),
            ToolInput(
                tag="selectRandomFraction",
                input_type=String(optional=True),
                prefix="--select-random-fraction",
                separate_value_from_prefix=True,
                doc="(-fraction:Double)  Select a fraction of variants at random from the input  Default value: 0.0. ",
            ),
            ToolInput(
                tag="selectTypeToExclude",
                input_type=String(optional=True),
                prefix="--select-type-to-exclude",
                separate_value_from_prefix=True,
                doc="(-xl-select-type:Type)  Do not select certain type of variants from the input file  This argument may be specified 0 or more times. Default value: null. Possible values: {NO_VARIATION, SNP, MNP, INDEL, SYMBOLIC, MIXED} ",
            ),
            ToolInput(
                tag="selectTypeToInclude",
                input_type=String(optional=True),
                prefix="--select-type-to-include",
                separate_value_from_prefix=True,
                doc="(-select-type:Type)  Select only a certain type of variants from the input file  This argument may be specified 0 or more times. Default value: null. Possible values: {NO_VARIATION, SNP, MNP, INDEL, SYMBOLIC, MIXED} ",
            ),
            ToolInput(
                tag="selectexpressions",
                input_type=String(optional=True),
                prefix="--selectExpressions",
                separate_value_from_prefix=True,
                doc="(-select:String)  One or more criteria to use when selecting the data  This argument may be specified 0 or more times. Default value: null. ",
            ),
            ToolInput(
                tag="sequenceDictionary",
                input_type=File(optional=True),
                prefix="-sequence-dictionary",
                separate_value_from_prefix=True,
                doc="(--sequence-dictionary)  Use the given sequence dictionary as the master/canonical sequence dictionary.  Must be a .dict file.  Default value: null. ",
            ),
            ToolInput(
                tag="setFilteredGtToNocall",
                input_type=Boolean(optional=True),
                prefix="--set-filtered-gt-to-nocall",
                separate_value_from_prefix=True,
                doc=" Set filtered genotypes to no-call  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="sitesOnlyVcfOutput",
                input_type=Boolean(optional=True),
                prefix="--sites-only-vcf-output",
                separate_value_from_prefix=True,
                doc=" If true, don't emit genotype fields when writing vcf file output.  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="tmpDir",
                input_type=Filename(optional=True),
                prefix="--tmp-dir",
                separate_value_from_prefix=True,
                doc="Temp directory to use. Default value: null.",
            ),
            ToolInput(
                tag="jdkDeflater",
                input_type=Boolean(optional=True),
                prefix="-jdk-deflater",
                separate_value_from_prefix=True,
                doc="(--use-jdk-deflater)  Whether to use the JdkDeflater (as opposed to IntelDeflater)  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="jdkInflater",
                input_type=Boolean(optional=True),
                prefix="-jdk-inflater",
                separate_value_from_prefix=True,
                doc="(--use-jdk-inflater)  Whether to use the JdkInflater (as opposed to IntelInflater)  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="verbosity",
                input_type=String(optional=True),
                prefix="-verbosity",
                separate_value_from_prefix=True,
                doc="(--verbosity)  Control verbosity of logging.  Default value: INFO. Possible values: {ERROR, WARNING, INFO, DEBUG} ",
            ),
            ToolInput(
                tag="version",
                input_type=Boolean(optional=True),
                prefix="--version",
                separate_value_from_prefix=True,
                doc="display the version number for this tool Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="disableToolDefaultReadFilters",
                input_type=Boolean(optional=True),
                prefix="-disable-tool-default-read-filters",
                separate_value_from_prefix=True,
                doc="(--disable-tool-default-read-filters)  Disable all tool default read filters (WARNING: many tools will not function correctly without their default read filters on)  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="showhidden",
                input_type=Boolean(optional=True),
                prefix="-showHidden",
                separate_value_from_prefix=True,
                doc="(--showHidden)  display hidden arguments  Default value: false. Possible values: {true, false} ",
            ),
            # Conditional Arguments for readFilter:
            ToolInput(
                tag="ambigFilterBases",
                input_type=Int(optional=True),
                prefix="--ambig-filter-bases",
                separate_value_from_prefix=True,
                doc="Valid only if 'AmbiguousBaseReadFilter' is specified: Threshold number of ambiguous bases. If null, uses threshold fraction; otherwise, overrides threshold fraction.  Default value: null.  Cannot be used in conjuction with argument(s) maxAmbiguousBaseFraction",
            ),
            ToolInput(
                tag="ambigFilterFrac",
                input_type=Double(optional=True),
                prefix="--ambig-filter-frac",
                separate_value_from_prefix=True,
                doc="Valid only if 'AmbiguousBaseReadFilter' is specified: Threshold fraction of ambiguous bases Default value: 0.05. Cannot be used in conjuction with argument(s) maxAmbiguousBases",
            ),
            ToolInput(
                tag="maxFragmentLength",
                input_type=Int(optional=True),
                prefix="--max-fragment-length",
                separate_value_from_prefix=True,
                doc="Valid only if 'FragmentLengthReadFilter' is specified: Maximum length of fragment (insert size) Default value: 1000000.",
            ),
            ToolInput(
                tag="minFragmentLength",
                input_type=Int(optional=True),
                prefix="--min-fragment-length",
                separate_value_from_prefix=True,
                doc="Valid only if 'FragmentLengthReadFilter' is specified: Minimum length of fragment (insert size) Default value: 0.",
            ),
            ToolInput(
                tag="keepIntervals",
                input_type=String(optional=True),
                prefix="--keep-intervals",
                separate_value_from_prefix=True,
                doc="Valid only if 'IntervalOverlapReadFilter' is specified: One or more genomic intervals to keep This argument must be specified at least once. Required. ",
            ),
            ToolInput(
                tag="library",
                input_type=String(optional=True),
                prefix="-library",
                separate_value_from_prefix=True,
                doc="Valid only if 'LibraryReadFilter' is specified: (--library) Name of the library to keep This argument must be specified at least once. Required.",
            ),
            ToolInput(
                tag="maximumMappingQuality",
                input_type=Int(optional=True),
                prefix="--maximum-mapping-quality",
                separate_value_from_prefix=True,
                doc="Valid only if 'MappingQualityReadFilter' is specified: Maximum mapping quality to keep (inclusive)  Default value: null. ",
            ),
            ToolInput(
                tag="minimumMappingQuality",
                input_type=Int(optional=True),
                prefix="--minimum-mapping-quality",
                separate_value_from_prefix=True,
                doc="Valid only if 'MappingQualityReadFilter' is specified: Minimum mapping quality to keep (inclusive)  Default value: 10. ",
            ),
            ToolInput(
                tag="dontRequireSoftClipsBothEnds",
                input_type=Boolean(optional=True),
                prefix="--dont-require-soft-clips-both-ends",
                separate_value_from_prefix=True,
                doc="Valid only if 'OverclippedReadFilter' is specified: Allow a read to be filtered out based on having only 1 soft-clipped block. By default, both ends must have a soft-clipped block, setting this flag requires only 1 soft-clipped block  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="filterTooShort",
                input_type=Int(optional=True),
                prefix="--filter-too-short",
                separate_value_from_prefix=True,
                doc="Valid only if 'OverclippedReadFilter' is specified: Minimum number of aligned bases Default value: 30.",
            ),
            ToolInput(
                tag="platformFilterName",
                input_type=String(optional=True),
                prefix="--platform-filter-name",
                separate_value_from_prefix=True,
                doc="Valid only if 'PlatformReadFilter' is specified: This argument must be specified at least once. Required.",
            ),
            ToolInput(
                tag="blackListedLanes",
                input_type=String(optional=True),
                prefix="--black-listed-lanes",
                separate_value_from_prefix=True,
                doc="Valid only if 'PlatformUnitReadFilter' is specified: Platform unit (PU) to filter out This argument must be specified at least once. Required.",
            ),
            ToolInput(
                tag="readGroupBlackList",
                input_type=String(optional=True),
                prefix="--read-group-black-list",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadGroupBlackListReadFilter' is specified: The name of the read group to filter out. This argument must be specified at least once. Required. ",
            ),
            ToolInput(
                tag="keepReadGroup",
                input_type=String(optional=True),
                prefix="--keep-read-group",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadGroupReadFilter' is specified: The name of the read group to keep Required.",
            ),
            ToolInput(
                tag="maxReadLength",
                input_type=Int(optional=True),
                prefix="--max-read-length",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadLengthReadFilter' is specified: Keep only reads with length at most equal to the specified value Required.",
            ),
            ToolInput(
                tag="minReadLength",
                input_type=Int(optional=True),
                prefix="--min-read-length",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadLengthReadFilter' is specified: Keep only reads with length at least equal to the specified value Default value: 1.",
            ),
            ToolInput(
                tag="readName",
                input_type=String(optional=True),
                prefix="--read-name",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadNameReadFilter' is specified: Keep only reads with this read name Required.",
            ),
            ToolInput(
                tag="keepReverseStrandOnly",
                input_type=Boolean(optional=True),
                prefix="--keep-reverse-strand-only",
                separate_value_from_prefix=True,
                doc="Valid only if 'ReadStrandFilter' is specified: Keep only reads on the reverse strand  Required. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="sample",
                input_type=String(optional=True),
                prefix="--sample",
                separate_value_from_prefix=True,
                doc="Valid only if 'SampleReadFilter' is specified: The name of the sample(s) to keep, filtering out all others This argument must be specified at least once. Required. ",
            ),
            ToolInput(
                tag="invertSoftClipRatioFilter",
                input_type=Boolean(optional=True),
                prefix="--invert-soft-clip-ratio-filter",
                separate_value_from_prefix=True,
                doc=" Inverts the results from this filter, causing all variants that would pass to fail and visa-versa.  Default value: false. Possible values: {true, false} ",
            ),
            ToolInput(
                tag="softClippedLeadingTrailingRatio",
                input_type=Double(optional=True),
                prefix="--soft-clipped-leading-trailing-ratio",
                separate_value_from_prefix=True,
                doc=" Threshold ratio of soft clipped bases (leading / trailing the cigar string) to total bases in read for read to be filtered.  Default value: null.  Cannot be used in conjuction with argument(s) minimumSoftClippedRatio",
            ),
            ToolInput(
                tag="softClippedRatioThreshold",
                input_type=Double(optional=True),
                prefix="--soft-clipped-ratio-threshold",
                separate_value_from_prefix=True,
                doc=" Threshold ratio of soft clipped bases (anywhere in the cigar string) to total bases in read for read to be filtered.  Default value: null.  Cannot be used in conjuction with argument(s) minimumLeadingTrailingSoftClippedRatio",
            ),
        ]

    def outputs(self):
        return [ToolOutput("out", VcfTabix(), glob=InputSelector("outputFilename"))]

    def metadata(self):
        return ToolMetadata(
            contributors=[],
            dateCreated=datetime.fromisoformat("2019-12-02T15:30:28.092132"),
            dateUpdated=datetime.fromisoformat("2019-12-02T15:30:28.092133"),
            documentation="USAGE: Selectvariants [arguments]\nThis tool makes it possible to select a subset of variants based on various criteria in order to facilitate certain\nanalyses. Examples include comparing and contrasting cases vs. controls, extracting variant or non-variant loci that\nmeet certain requirements, or troubleshooting some unexpected results, to name a few.\nVersion:4.1.3.0\n",
        )
