from abc import ABC
from datetime import datetime

from janis_bioinformatics.tools.bioinformaticstoolbase import BioinformaticsTool
from janis_unix import Csv, Tsv

from janis_bioinformatics.data_types import Bam, Fasta
from janis_core import (
    CommandTool,
    ToolInput,
    ToolOutput,
    File,
    Boolean,
    String,
    Int,
    Double,
    Float,
    InputSelector,
    Filename,
    ToolMetadata,
    InputDocumentation,
    Array,
)


class ArribaBase(BioinformaticsTool, ABC):
    def friendly_name(self) -> str:
        return "Arriba"

    def tool_provider(self):
        return "Arriba"

    def tool(self) -> str:
        return "Arriba"

    def base_command(self):
        return ["arriba"]

    def inputs(self):
        return [
            ToolInput(
                tag="aligned_inp",
                input_type=Bam(),
                prefix="-x",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="File in SAM/BAM/CRAM format with main alignments as generated by STAR (Aligned.out.sam). "
                    "Arriba extracts candidate reads from this file. This is sometimes /dev/stdin"
                ),
            ),
            ToolInput(
                tag="inp_chimeric",
                input_type=Bam(optional=True),
                prefix="-c",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="File in SAM/BAM/CRAM format with chimeric alignments as generated by STAR (Chimeric.out.sam). "
                    "This parameter is only required, if STAR was run with the parameter "
                    "'--chimOutType SeparateSAMold'. When STAR was run with the parameter "
                    "'--chimOutType WithinBAM', it suffices to pass the parameter -x to Arriba and -c can be omitted. "
                ),
            ),
            ToolInput(
                tag="gtf_file",
                input_type=File(optional=True),
                prefix="-g",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="GTF file with gene annotation. The file may be gzip-compressed."
                ),
            ),
            ToolInput(
                tag="gtf_features",
                input_type=Csv(optional=True),
                prefix="-G",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Comma-/space-separated list of names of GTF features. "
                    "Default: gene_name=gene_name|gene_id gene_id=gene_id transcript_id=transcript_id feature_exon=exon feature_CDS=CDS "
                ),
            ),
            ToolInput(
                tag="reference",
                input_type=Fasta(optional=True),
                prefix="-a",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="FastA file with genome sequence (assembly). The file may be gzip-compressed. An index with "
                    "the file extension .fai must exist only if CRAM files are processed. "
                ),
            ),
            ToolInput(
                tag="blacklist",
                input_type=File(optional=True),
                prefix="-b",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="File containing blacklisted events (recurrent artifacts and transcripts observed in healthy tissue). "
                ),
            ),
            ToolInput(
                tag="known_fusions",
                input_type=Tsv(optional=True),
                prefix="-k",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="File containing known/recurrent fusions. Some cancer entities are often characterized by "
                    "fusions between the same pair of genes. In order to boost sensitivity, a list of known "
                    "fusions can be supplied using this parameter. The list must contain two columns with the "
                    "names of the fused genes, separated by tabs. "
                ),
            ),
            ToolInput(
                tag="output_filename",
                input_type=Filename(extension=".tsv"),
                prefix="-o",
                default="fusions.tsv",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Output file with fusions that have passed all filters."
                ),
            ),
            ToolInput(
                tag="discarded_output_filename",
                input_type=Filename(suffix=".discarded", extension=".tsv"),
                prefix="-O",
                separate_value_from_prefix=True,
                default="fusions.discarded.tsv",
                doc=InputDocumentation(
                    doc="Output file with fusions that were discarded due to filtering."
                ),
            ),
            ToolInput(
                tag="structural_variants_coordinates",
                input_type=Tsv(optional=True),
                prefix="-d",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Tab-separated file with coordinates of structural variants found using whole-genome "
                    "sequencing data. These coordinates serve to increase sensitivity towards weakly expressed "
                    "fusions and to eliminate fusions with low evidence. "
                ),
            ),
            ToolInput(
                tag="max_genomic_breakpoint_distance",
                input_type=Int(optional=True),
                prefix="-D",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When a file with genomic breakpoints obtained via whole-genome sequencing is supplied via "
                    "the -d parameter, this parameter determines how far a genomic breakpoint may be away from a "
                    "transcriptomic breakpoint to consider it as a related event. For events inside genes, the "
                    "distance is added to the end of the gene; for intergenic events, the distance threshold is "
                    "applied as is. Default: 100000 "
                ),
            ),
            ToolInput(
                tag="strandedness",
                input_type=String(optional=True),
                prefix="-s",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Whether a strand-specific protocol was used for library preparation, and if so, the type of "
                    "strandedness (auto/yes/no/reverse). When unstranded data is processed, the strand can "
                    "sometimes be inferred from splice-patterns. But in unclear situations, stranded data helps"
                    " resolve ambiguities. Default: auto "
                ),
            ),
            ToolInput(
                tag="contigs",
                input_type=Array(String(), optional=True),
                prefix="-i",
                doc=InputDocumentation(
                    doc="Comma-/space-separated list of interesting contigs. Fusions between genes on other contigs "
                    "are ignored. Contigs can be specified with or without the prefix 'chr'. "
                    "Default: 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y "
                ),
            ),
            ToolInput(
                tag="filters",
                input_type=Array(String, optional=True),
                prefix="-f",
                separator=" ",
                doc=InputDocumentation(
                    doc="Comma-/space-separated list of filters to disable. By default all filters are enabled. "
                    "Valid values: homopolymer, same_gene, inconsistently_clipped, duplicates, low_entropy, "
                    "no_genomic_support, short_anchor, homologs, blacklist, pcr_fusions, isoforms, intronic, "
                    "uninteresting_contigs, read_through, genomic_support, mismatches, no_coverage, spliced, "
                    "mismappers, merge_adjacent, select_best, many_spliced, long_gap, min_support, "
                    "relative_support, end_to_end, known_fusions, non_coding_neighbors, intragenic_exonic, "
                    "hairpin, small_insert_size "
                ),
            ),
            ToolInput(
                tag="max_e_value",
                input_type=Float(optional=True),
                prefix="-E",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Arriba estimates the number of fusions with a given number of supporting reads which one "
                    "would expect to see by random chance. If the expected number of fusions (e-value) is higher "
                    "than this threshold, the fusion is discarded by the 'relative_support' filter. Note: "
                    "Increasing this threshold can dramatically increase the number of false positives and may "
                    "increase the runtime of resource-intensive steps. Fractional values are possible. "
                    "Default: 0.300000 "
                ),
            ),
            ToolInput(
                tag="min_supporting_reads",
                input_type=Int(optional=True),
                prefix="-S",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'min_support' filter discards all fusions with fewer than this many supporting reads "
                    "(split reads and discordant mates combined). Default: 2 "
                ),
            ),
            ToolInput(
                tag="max_mismappers",
                input_type=Float(optional=True),
                prefix="-m",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When more than this fraction of supporting reads turns out to be mismappers, the "
                    "'mismappers' filter discards the fusion. Default: 0.800000 "
                ),
            ),
            ToolInput(
                tag="max_homolog_identity",
                input_type=Float(optional=True),
                prefix="-L",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Genes with more than the given fraction of sequence identity are considered homologs and "
                    "removed by the 'homologs' filter. Default: 0.300000 "
                ),
            ),
            ToolInput(
                tag="homopolymer_length",
                input_type=Int(optional=True),
                prefix="-H",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'homopolymer' filter removes breakpoints adjacent to homopolymers of the given length "
                    "or more. Default: 6 "
                ),
            ),
            ToolInput(
                tag="read_through_distance",
                input_type=Int(optional=True),
                prefix="-R",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'read_through' filter removes read-through fusions where the breakpoints are "
                    "less than the given distance away from each other. Default: 10000 "
                ),
            ),
            ToolInput(
                tag="min_anchor_length",
                input_type=Int(optional=True),
                prefix="-A",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Alignment artifacts are often characterized by split reads coming from only one gene "
                    "and no discordant mates. Moreover, the split reads only align to a short stretch in one "
                    "of the genes. The 'short_anchor' filter removes these fusions. This parameter sets the "
                    "threshold in bp for what the filter considers short. Default: 23 "
                ),
            ),
            ToolInput(
                tag="many_spliced_events",
                input_type=Int(optional=True),
                prefix="-M",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'many_spliced' filter recovers fusions between genes that have at least this "
                    "many spliced breakpoints. Default: 4 "
                ),
            ),
            ToolInput(
                tag="max_kmer_content",
                input_type=Float(optional=True),
                prefix="-K",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'low_entropy' filter removes reads with repetitive 3-mers. If the 3-mers make up more "
                    "than the given fraction of the sequence, then the read is discarded. Default: 0.600000 "
                ),
            ),
            ToolInput(
                tag="max_mismatch_pvalue",
                input_type=Float(optional=True),
                prefix="-V",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The 'mismatches' filter uses a binomial model to calculate a p-value for observing a given "
                    "number of mismatches in a read. If the number of mismatches is too high, the read is "
                    "discarded. Default: 0.010000 "
                ),
            ),
            ToolInput(
                tag="fragment_length",
                input_type=Int(optional=True),
                prefix="-F",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When paired-end data is given, the fragment length is estimated automatically and this "
                    "parameter has no effect. But when single-end data is given, the mean fragment length "
                    "should be specified to effectively filter fusions that arise from hairpin structures. "
                    "Default: 200 "
                ),
            ),
            ToolInput(
                tag="max_reads",
                input_type=Int(optional=True),
                prefix="-U",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Subsample fusions with more than the given number of supporting reads. This improves "
                    "performance without compromising sensitivity, as long as the threshold is high. Counting "
                    "of supporting reads beyond the threshold is inaccurate, obviously. Default: 300 "
                ),
            ),
            ToolInput(
                tag="quantile",
                input_type=Float(optional=True),
                prefix="-Q",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="Highly expressed genes are prone to produce artifacts during library preparation. Genes "
                    "with an expression above the given quantile are eligible for filtering by the 'pcr_fusions' "
                    "filter. Default: 0.998000 "
                ),
            ),
            ToolInput(
                tag="exonic_fraction",
                input_type=Float(optional=True),
                prefix="-e",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="The breakpoints of false-positive predictions of intragenic events are often both in exons. "
                    "True predictions are more likely to have at least one breakpoint in an intron, because "
                    "introns are larger. If the fraction of exonic sequence between two breakpoints is smaller "
                    "than the given fraction, the 'intragenic_exonic' filter discards the event. Default: 0.200000"
                ),
            ),
            ToolInput(
                tag="fusion_transcript",
                input_type=Boolean(optional=True),
                prefix="-T",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When set, the column 'fusion_transcript' is populated with the sequence of the fused genes "
                    "as assembled from the supporting reads. Specify the flag twice to also print the fusion "
                    "transcripts to the file containing discarded fusions (-O). Default: off "
                ),
            ),
            ToolInput(
                tag="peptide_sequence",
                input_type=Boolean(optional=True),
                prefix="-P",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When set, the column 'peptide_sequence' is populated with the sequence of the fused proteins "
                    "as assembled from the supporting reads. Specify the flag twice to also print the peptide "
                    "sequence to the file containing discarded fusions (-O). Default: off "
                ),
            ),
            ToolInput(
                tag="read_identifiers",
                input_type=Boolean(optional=True),
                prefix="-I",
                separate_value_from_prefix=True,
                doc=InputDocumentation(
                    doc="When set, the column 'read_identifiers' is populated with identifiers of the reads which "
                    "support the fusion. The identifiers are separated by commas. Specify the flag twice to "
                    "also print the read identifiers to the file containing discarded fusions (-O). Default: off "
                ),
            ),
            # ToolInput(
            #   tag="help",
            #   input_type=Boolean(optional=True),
            #   prefix="-h",
            #   separate_value_from_prefix=True,
            #   doc=InputDocumentation(doc="Print help and exit."),
            # ),
        ]

    def outputs(self):
        return [
            ToolOutput("out", Tsv, selector=InputSelector("output_filename")),
            ToolOutput(
                "out_discarded",
                Tsv,
                selector=InputSelector("discarded_output_filename"),
            ),
        ]

    def bind_metadata(self):
        return ToolMetadata(
            contributors=["Michael Franklin"],
            dateCreated=datetime(2020, 9, 2),
            dateUpdated=datetime(2020, 9, 2),
            documentation="""
Arriba gene fusion detector
--------------------------- 
Version: 1.2.0 
Arriba is a fast tool to search for aberrant transcripts such as gene fusions.  
It is based on chimeric alignments found by the STAR RNA-Seq aligner. 

Arriba is a command-line tool for the detection of gene fusions from RNA-Seq data. It was developed for the use in a 
clinical research setting. Therefore, short runtimes and high sensitivity were important design criteria. It is based 
on the ultrafast STAR aligner and the post-alignment runtime is typically just ~2 minutes. In contrast to many other 
fusion detection tools which build on STAR, Arriba does not require to reduce the alignIntronMax parameter of STAR 
to detect fusions arising from focal deletions.

Apart from gene fusions, Arriba can detect other structural rearrangements with potential clinical relevance, such 
as exon duplications or truncations of genes (i.e., breakpoints in introns and intergenic regions).
""",
        )
